external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala

```diff
@@ -166,35 +204,91 @@ private[sql] class AvroSerializer(
         (getter, ordinal) => ByteBuffer.wrap(getter.getBinary(ordinal))
 
       case (DateType, INT) =>
-        (getter, ordinal) => dateRebaseFunc(getter.getInt(ordinal))
+        (getter, ordinal) =>
+          getter.get(ordinal, DateType) match {
+            case epochDays: java.lang.Integer => dateRebaseFunc(epochDays)
+            case date: java.sql.Date => dateRebaseFunc(date.toLocalDate().toEpochDay().toInt)
+            case localDate: java.time.LocalDate => dateRebaseFunc(localDate.toEpochDay().toInt)
+            case other =>
+              throw new IncompatibleSchemaException(s"""
+                  |Expected java.lang.Integer, java.sql.Date, or java.time.LocalDate,
+                  | but found ${other.getClass}""".stripMargin)
+          }
 
-      case (TimestampType, LONG) => avroType.getLogicalType match {
+      case (TimestampType, LONG) =>
+        avroType.getLogicalType match {
           // For backward compatibility, if the Avro type is Long and it is not logical type
           // (the `null` case), output the timestamp value as with millisecond precision.
-          case null | _: TimestampMillis => (getter, ordinal) =>
-            DateTimeUtils.microsToMillis(timestampRebaseFunc(getter.getLong(ordinal)))
-          case _: TimestampMicros => (getter, ordinal) =>
-            timestampRebaseFunc(getter.getLong(ordinal))
-          case other => throw new IncompatibleSchemaException(errorPrefix +
-            s"SQL type ${TimestampType.sql} cannot be converted to Avro logical type $other")
+          case null | _: TimestampMillis =>
+            (getter, ordinal) =>
+              getter.get(ordinal, TimestampType) match {
+                case micros: java.lang.Long =>
+                  DateTimeUtils.microsToMillis(timestampRebaseFunc(micros))
+                case javaTimestamp: java.sql.Timestamp => javaTimestamp.getTime
+                case instant: java.time.Instant => instant.toEpochMilli
+                case other =>
```

Comment:
I realized that it is not as easy as the other case. 
We can refactor out an `extractMillis` or `extractMicros` method, but it will introduce some computational overhead. There would always be a case of unnecessary back and forth conversion.